File: ria/java/gof/src/main/java/com/gof/pattern/singleton/Settings3.java
package com.gof.pattern.singleton;

public class Settings3 {

    /** 이른 초기화 사용하기 -> 멀티스레드 환경에서 안전함 / 비용 이슈 존재 **/
    private static final Settings3 INSTANCE = new Settings3();

    private Settings3() {
    }

    public static Settings3 getInstance() {
        return INSTANCE;
    }
}

File: zeri/gof/src/main/kotlin/com/gof/pattern/singleton/javakotlin/Settings3.kt
package com.gof.pattern.singleton.javakotlin

/**
 * [멀티 쓰레드 환경에서 안전하게 구현하는 방법]
 * - synchronized 사용하기
 */
class Settings3 private constructor() {
    companion object {
        private var instance: Settings3? = null

        @Synchronized
        fun getInstance(): Settings3 {
            return instance ?: Settings3().also { instance = it }
        }
    }
}
